proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,153 @@
+  application_manager->resume_controller().do_staff();
+  
+}
+```
+
+#### New Approach 
+```cpp
+
+SomeCommand::SomeCommand(protocol_handler, state_controller, policy_handler, resume_controller);
+
+void SomeCommand::Run() {
+  protocol_handler_.do_staff();
+  state_controller_.do_staff();
+  protocol_handler_.do_staff();
+  protocol_handler_.do_staff();
+}
+```
+This will simplify code and provide possibility to add a new component to certain commands without affecting other commands.
+And factory will create command with all required parameters, so existing code will not be complicated.  
+RC component contains class ResourceAllocationManager. And non-RC commands should have no access to this class, so it should be passed only to RC commands. 
+
+### Reduce responsibility of ApplicationManager class 
+
+Logic of handling and sending RPC's should be extracted from ApplicationManager to separate components `RPCHandler`, `RPCService`.
+
+### RPCHandler
+RPCHandler class should implement interfaces ```HMIMessageObserver```, ```ProtocolObserver```, handle incoming RPC's from HMI or mobile 
+and process them. 
+
+RPCHandler responsibilities:
+ - Handle HMI/Mobile RPC's in message loop
+ - Class checking policies on each RPC
+ - Creating command from message with Command Factories
+ - Send command for execution to Request Controller
+ 
+ ### RPCService
+
+RPCService should provide interface for sending messages to Mobile/HMI.  
+RPCService should be passed to commands for Sending RPC's using dependency injection approach.
+
+RPCService responsibilities:
+ - SendMessage to HMI/Mobile
+ - Call checking policies if message is allowed to send
+ - Provide Request Controller information of any incoming request should be terminated.  
+
+### Provide plugin architecture for easy adding extra functionality and RPC groups   
+
+Create **PluginManager** class that will be responsible for searching and loading rpc plugins.
+All commands logic should be stored in rpc plugins. 
+
+Each RPC plugin should implement following interfaces:
+ - Command Interface :
+ - CommandFactory Interface : 
+
+#### Plugins responsibilities:
+ - Provide list of supported commands
+ - Create commands from smart object
+ - Contain command logic
+
+SDL will provide to *Plugin* interfaces of all services on application manager layer:
+ - RPCService
+ - PolicyHander
+ - StateController
+ - ApplicationManager
+ - RequestController
+ - ResumeController
+ - etc ...
+
+RPC Handler will use PluginManager to get plugins that is able to process required RPC. 
+RPC Handler will use Command Factory from plugin to create Command for further processing.
+
+![New design approach](../assets/proposals/nnnn-rpc-design-refactoring/new_design.png)
+
+## Potential downsides
+This change requires major changes of application manager layer. 
+## Impact on existing code
+
+## Alternatives considered
+The only alternatives is to following existing approach, duplicating logic of application_manager layer. 
```

Comment:
The only alternatives is to following existing approach, duplicating the logic of application_manager layer -- this approach is becoming unsustainable and should be avoided.

Reply:
@mrapitis, done in 2c9f7a3f7bb7e6c8b26bb1e4aec779cf65733666